lib/astutils.cpp

```diff
@@ -133,9 +133,74 @@ const Token * astIsVariableComparison(const Token *tok, const std::string &comp,
     return ret;
 }
 
+const Token * getVariableInitExpression(const Variable * var)
+{
+    if(!var || !var->declEndToken())
+        return nullptr;
+    if(Token::Match(var->declEndToken(), "; %varid% =", var->declarationId()))
+        return var->declEndToken()->tokAt(2)->astOperand2();
+    return var->declEndToken()->astOperand2();
+}
+
+bool isInLoopCondition(const Token * tok) {
+    return Token::Match(tok->astTop()->previous(), "for|while (");
+}
+
+
+/// This takes a token that refers to a variable and it will return the token
+/// to the expression that the variable is assigned to. If its not valid to
+/// make such substitution then it will return the original token.
+const Token * followVariableExpression(const Token * tok, bool cpp)
+{
+    if(!tok)
+        return tok;
+    // Skip array access
+    if(Token::Match(tok, "%var% ["))
+        return tok;
+    // Skip following variables if it is used in an assignment
+    if(Token::Match(tok->astParent(), "%assign%") || Token::Match(tok->next(), "%assign%"))
+        return tok;
+    const Variable * var = tok->variable();
+    const Token * varTok = getVariableInitExpression(var);
+    if(!varTok)
+        return tok;
+    // Skip array access
+    if(Token::simpleMatch(varTok, "["))
+        return tok;
+    if(var->scope() != tok->scope() && !var->isConst())
+        return tok;
+    if(var->isStatic() && !var->isConst())
+        return tok;
+    if(var->isArgument())
+        return tok;
+    // If this is in a loop then check if variables are modified in the entire scope
+    const Token * endToken = isInLoopCondition(tok) ? tok->scope()->bodyEnd : tok;
+    const Token * startToken = varTok;
+    while(Token::Match(startToken, "%op%|.|(|{"))
+        startToken = startToken->astOperand1();
+    // Skip if the variable its referring to is modified
+    for(const Token * tok2 = startToken;tok2 != endToken;tok2 = tok2->next()) {
```

Comment:
it seems to me that you only want to parse the initialization expression here. It is a suggestion, I would have used the AST. Like:

    std::stack<const Token *> tokens;
    tokens.push(startToken);
    while (!tokens.empty()) {
        const Token *tok = tokens.top();
        tokens.pop();
        if (!tok)
            continue;
        tokens.push(tok->astOperand1());
        tokens.push(tok->astOperand2());
        .....
    }

However I agree that we get more code with this approach. :-(

We have such loops in couple of places, it's unfortunate, maybe we can create some helper function someday.

Reply:
If we are able to use lambdas in the source code we can create a higher-order function to do the ast walking:

```cpp
template<class F>
void walk(const Token* startToken, F f) {
    std::stack<const Token *> tokens;
    tokens.push(startToken);
    while (!tokens.empty()) {
        const Token *tok = tokens.top();
        f(tok);
        tokens.pop();
        if (!tok)
            continue;
        tokens.push(tok->astOperand1());
        tokens.push(tok->astOperand2());
    }
}
```

Then use it like:

```cpp
walk(startToken, [&](const Token * tok) {
    ...
});
```

Reply:
Nice suggestion. That should be ok. I will look into it a little.